### PR TITLE
Use current Ogre3D master (ca665a6) for rviz_ogre_vendor

### PR DESCRIPTION
**Pull request by anhosi**
- This fixes rendering problems the point cloud (now all styles are
  working)
- Ogre should be updated to their next release when available. We use
  the current master only because of the upcoming ros2 release.

The Ogre::SceneManager constructor with a single argument has been deprecated.

### CODE REVIEW

**Comment by wjwwood** [comment]
I don't think we can get rid of the patch, it is still required on macOS I believe. If someone can test on macOS then I can go along with it, but right now I don't have the cycles free to test it out.

**Comment by anhosi** [comment]
We successfully tested this pr on Windows, Linux, and MacOSX. The extra patch is no longer required.

File: rviz_rendering/CMakeLists.txt [hunk 1]
```diff
@@ -175,4 +175,6 @@ list(APPEND ${PROJECT_NAME}_CONFIG_EXTRAS
   "${CMAKE_CURRENT_SOURCE_DIR}/src/cmake/register_rviz_ogre_media_exports_hook-extras.cmake"
   "${CMAKE_CURRENT_SOURCE_DIR}/src/cmake/register_rviz_ogre_media_exports.cmake")
 
+set_target_properties(rendering_example PROPERTIES INSTALL_RPATH_USE_LINK_PATH TRUE)
```
Comment (anhosi) [inline rviz_rendering/CMakeLists.txt:178]: @wjwwood We need this for the linker to succeed. I remember that you do not need this. We are not quite sure what is causing this different behavior.

**Comment by wjwwood** [comment]
Hmm, my `rviz_rendering` is unable to find package Ogre after this pr:

```
Process package 'rviz_rendering' with context:
--------------------------------------------------------------------------------
 source_space => /Users/william/rviz2_ws/src/rviz/rviz_rendering
  build_space => /Users/william/rviz2_ws/build/rviz_rendering
install_space => /Users/william/rviz2_ws/install
   make_flags => -j4, -l4
  build_tests => True
--------------------------------------------------------------------------------
+++ Building 'rviz_rendering'
==> '. /Users/william/rviz2_ws/build/rviz_rendering/cmake__build.sh && /usr/bin/make -j4 -l4' in '/Users/william/rviz2_ws/build/rviz_rendering'
-- Found ament_cmake: 0.0.3 (/Users/william/ros2_ws/install/share/ament_cmake/cmake)
-- Using PYTHON_EXECUTABLE: /usr/local/bin/python3
-- Override CMake install command with custom implementation using symlinks instead of copying resources
-- Found ament_cmake_gtest: 0.0.3 (/Users/william/ros2_ws/install/share/ament_cmake_gtest/cmake)
-- Found ament_index_cpp: 0.0.3 (/Users/william/ros2_ws/install/share/ament_index_cpp/cmake)
-- Found rviz_ogre_vendor: 1.0.0 (/Users/william/rviz2_ws/install/share/rviz_ogre_vendor/cmake)
-- Setting OGRE_DIR to: '/Users/william/rviz2_ws/install/share/rviz_ogre_vendor/cmake/../../../opt/rviz_ogre_vendor/lib/OGRE/cmake'
CMake Error at /Users/william/rviz2_ws/install/share/rviz_ogre_vendor/cmake/rviz_ogre_vendorConfig.cmake:26 (find_package):
  By not providing "FindOGRE.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "OGRE", but
  CMake did not find one.

  Could not find a package configuration file provided by "OGRE" with any of
  the following names:

    OGREConfig.cmake
    ogre-config.cmake

  Add the installation prefix of "OGRE" to CMAKE_PREFIX_PATH or set
  "OGRE_DIR" to a directory containing one of the above files.  If "OGRE"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  CMakeLists.txt:32 (find_package)


-- Configuring incomplete, errors occurred!
See also "/Users/william/rviz2_ws/build/rviz_rendering/CMakeFiles/CMakeOutput.log".
make: *** [cmake_check_build_system] Error 1

<== Command '. /Users/william/rviz2_ws/build/rviz_rendering/cmake__build.sh && /usr/bin/make -j4 -l4' failed in '/Users/william/rviz2_ws/build/rviz_rendering' with exit code '2'
<== Command '. /Users/william/rviz2_ws/build/rviz_rendering/cmake__build.sh && /usr/bin/make -j4 -l4' failed in '/Users/william/rviz2_ws/build/rviz_rendering' with exit code '2'
```

Part of the patch file was fixing some of those issues IIRC.

File: rviz_rendering/CMakeLists.txt (hunk 1 above)
Comment (wjwwood) [reply rviz_rendering/CMakeLists.txt:178]: I've actually got `rendering_example` disabled in the use pluginlib branch, maybe we should just get rid of the `rendering_example`.

**Comment by wjwwood** [comment]
@anhosi unfortunately this doesn't appear to fix the issue on my mac:

```
% lldb rviz2_app
(lldb) target create "rviz2_app"
Current executable set to 'rviz2_app' (x86_64).
(lldb) r
Process 20208 launched: '/Users/william/rviz2_ws/install/bin/rviz2_app' (x86_64)
2017-12-04 13:29:17.015819-0800 rviz2_app[20208:65518539] Month 13 is out of bounds
2017-12-04 13:29:17.058259-0800 rviz2_app[20208:65518539] Month 13 is out of bounds
2017-12-04 13:29:17.085956-0800 rviz2_app[20208:65518563] Month 13 is out of bounds
QPainter::begin: Paint device returned engine == 0, type: 2
QPainter::setPen: Painter not active
2017-12-04 13:29:17.679735-0800 rviz2_app[20208:65518539] Month 13 is out of bounds
2017-12-04 13:29:17.710024-0800 rviz2_app[20208:65518539] Month 13 is out of bounds
2017-12-04 13:29:17.710647-0800 rviz2_app[20208:65518539] Month 13 is out of bounds
2017-12-04 13:29:17.719509-0800 rviz2_app[20208:65518539] Month 13 is out of bounds
[INFO] [rviz2]: Creating resource group General
[INFO] [rviz2]: Creating resource group OgreInternal
[INFO] [rviz2]: Creating resource group OgreAutodetect
[INFO] [rviz2]: SceneManagerFactory for type 'DefaultSceneManager' registered.
[INFO] [rviz2]: Registering ResourceManager for type Material
[INFO] [rviz2]: Registering ResourceManager for type Mesh
[INFO] [rviz2]: Registering ResourceManager for type Skeleton
[INFO] [rviz2]: MovableObjectFactory for type 'ParticleSystem' registered.
[INFO] [rviz2]: ArchiveFactory for archive type FileSystem registered.
[INFO] [rviz2]: DDS codec registering
[INFO] [rviz2]: FreeImage version: 3.17.0
[INFO] [rviz2]: This program uses FreeImage, a free, open source image library supporting all common bitmap formats. See http://freeimage.sourceforge.net for details
[INFO] [rviz2]: Supported formats: bmp,ico,jpg,jif,jpeg,jpe,jng,koa,iff,lbm,mng,pbm,pbm,pcd,pcx,pgm,pgm,png,ppm,ppm,ras,tga,targa,tif,tiff,wap,wbmp,wbm,psd,cut,xbm,xpm,gif,hdr,g3,sgi,rgb,rgba,bw,exr,j2k,j2c,jp2,pfm,pct,pict,pic,3fr,arw,bay,bmq,cap,cine,cr2,crw,cs1,dc2,dcr,drf,dsc,dng,erf,fff,ia,iiq,k25,kc2,kdc,mdc,mef,mos,mrw,nef,nrw,orf,pef,ptx,pxn,qtk,raf,raw,rdc,rw2,rwl,rwz,sr2,srf,srw,sti,x3f,webp,jxr,wdp,hdp
[INFO] [rviz2]: Registering ResourceManager for type HighLevelGpuProgram
[INFO] [rviz2]: Registering ResourceManager for type Compositor
[INFO] [rviz2]: MovableObjectFactory for type 'Entity' registered.
[INFO] [rviz2]: MovableObjectFactory for type 'Light' registered.
[INFO] [rviz2]: MovableObjectFactory for type 'BillboardSet' registered.
[INFO] [rviz2]: MovableObjectFactory for type 'ManualObject' registered.
[INFO] [rviz2]: MovableObjectFactory for type 'BillboardChain' registered.
[INFO] [rviz2]: MovableObjectFactory for type 'RibbonTrail' registered.
[INFO] [rviz2]: *-*-* OGRE Initialising
[INFO] [rviz2]: *-*-* Version 1.10.10 (Xalafu)
[INFO] [rviz2]: OverlayElementFactory for type Panel registered.
[INFO] [rviz2]: OverlayElementFactory for type BorderPanel registered.
[INFO] [rviz2]: OverlayElementFactory for type TextArea registered.
[INFO] [rviz2]: Registering ResourceManager for type Font
[INFO] [rviz2]: Installing plugin: GL RenderSystem
[INFO] [rviz2]: OpenGL Rendering Subsystem created.
[INFO] [rviz2]: Plugin successfully installed
[INFO] [rviz2]: CPU Identifier & Features
[INFO] [rviz2]: -------------------------
[INFO] [rviz2]:  *   CPU ID: GenuineIntel: Intel(R) Core(TM) i7-7567U CPU @ 3.50GHz
[INFO] [rviz2]:  *          SSE: yes
[INFO] [rviz2]:  *         SSE2: yes
[INFO] [rviz2]:  *         SSE3: yes
[INFO] [rviz2]:  *        SSE41: yes
[INFO] [rviz2]:  *        SSE42: yes
[INFO] [rviz2]:  *          MMX: yes
[INFO] [rviz2]:  *       MMXEXT: yes
[INFO] [rviz2]:  *        3DNOW: no
[INFO] [rviz2]:  *     3DNOWEXT: no
[INFO] [rviz2]:  *         CMOV: yes
[INFO] [rviz2]:  *          TSC: yes
[INFO] [rviz2]:  *INVARIANT TSC: yes
[INFO] [rviz2]:  *          FPU: yes
[INFO] [rviz2]:  *          PRO: yes
[INFO] [rviz2]:  *           HT: no
[INFO] [rviz2]: -------------------------
[INFO] [rviz2]: ********************************************
***  Starting Mac OS X OpenGL Subsystem  ***
********************************************
[INFO] [rviz2]: Registering ResourceManager for type Texture
[INFO] [rviz2]: GLRenderSystem::_createRenderWindow "OgreWindow(0)", 1x1 windowed  miscParams: FSAA=4 border=none contentScalingFactor=1.000000 currentGLContext=false externalWindowHandle=0 left=0 macAPI=cocoa macAPICocoaUseNSView=true parentWindowHandle=0 top=0
[INFO] [rviz2]: Creating a Cocoa Compatible Render System
Assertion failed: ([externalHandle isKindOfClass:[NSView class]]), function create, file /Users/william/rviz2_ws/build/rviz_ogre_vendor/ogre-master-ca665a6-prefix/src/ogre-master-ca665a6/RenderSystems/GLSupport/src/OSX/OgreOSXCocoaWindow.mm, line 334.
Process 20208 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff695b5fce libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff695b5fce <+10>: jae    0x7fff695b5fd8            ; <+20>
    0x7fff695b5fd0 <+12>: movq   %rax, %rdi
    0x7fff695b5fd3 <+15>: jmp    0x7fff695ad76c            ; cerror_nocancel
    0x7fff695b5fd8 <+20>: retq
Target 0: (rviz2_app) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff695b5fce libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff696f3150 libsystem_pthread.dylib`pthread_kill + 333
    frame #2: 0x00007fff6951232a libsystem_c.dylib`abort + 127
    frame #3: 0x00007fff694da380 libsystem_c.dylib`__assert_rtn + 320
    frame #4: 0x000000010207d126 librviz_rendering.dylib`Ogre::CocoaWindow::create(this=0x00000001042a44e0, name="OgreWindow(0)", widthPt=1, heightPt=1, fullScreen=false, miscParams=0x00007ffeefbfd568 size=10) at OgreOSXCocoaWindow.mm:334
    frame #5: 0x000000010208ee2b librviz_rendering.dylib`Ogre::OSXGLSupport::newWindow(this=0x000000010d0b2600, name="OgreWindow(0)", width=1, height=1, fullScreen=false, miscParams=0x00007ffeefbfd568 size=10) at OgreOSXGL3PlusSupport.mm:312
    frame #6: 0x000000010202586c librviz_rendering.dylib`Ogre::GLSupport::newWindow(this=0x000000010d0b2a10, name="OgreWindow(0)", width=1, height=1, fullScreen=false, miscParams=0x00007ffeefbfd568 size=10) at OgreGLSupport.h:81
    frame #7: 0x0000000102024b46 librviz_rendering.dylib`Ogre::GLRenderSystem::_createRenderWindow(this=0x0000000105808c00, name="OgreWindow(0)", width=1, height=1, fullScreen=false, miscParams=0x00007ffeefbfd568 size=10) at OgreGLRenderSystem.cpp:1085
    frame #8: 0x0000000101c79867 librviz_rendering.dylib`Ogre::Root::createRenderWindow(this=0x00000001042a23a0, name="OgreWindow(0)", width=1, height=1, fullScreen=false, miscParams=0x00007ffeefbfd568 size=10) at OgreRoot.cpp:1249
    frame #9: 0x00000001016c6a0e librviz_rendering.dylib`rviz_rendering::RenderSystem::tryMakeRenderWindow(this=0x00000001042307c0, name="OgreWindow(0)", width=1, height=1, params=0x00007ffeefbfd568 size=10, max_attempts=100) at render_system.cpp:511
    frame #10: 0x00000001016c18fb librviz_rendering.dylib`rviz_rendering::RenderSystem::makeRenderWindow(this=0x00000001042307c0, window_id=0, width=1, height=1, pixel_ratio=1) at render_system.cpp:470
    frame #11: 0x00000001016bc701 librviz_rendering.dylib`rviz_rendering::RenderSystem::RenderSystem(this=0x00000001042307c0) at render_system.cpp:143
    frame #12: 0x00000001016bb6b5 librviz_rendering.dylib`rviz_rendering::RenderSystem::RenderSystem(this=0x00000001042307c0) at render_system.cpp:131
    frame #13: 0x00000001016bb65d librviz_rendering.dylib`rviz_rendering::RenderSystem::get() at render_system.cpp:79
    frame #14: 0x00000001016ce950 librviz_rendering.dylib`rviz_rendering::RenderWindowImpl::RenderWindowImpl(this=0x000000010429f680, parent=0x00000001042a16f0) at ogre_render_window_impl.cpp:96
    frame #15: 0x00000001016ce97d librviz_rendering.dylib`rviz_rendering::RenderWindowImpl::RenderWindowImpl(this=0x000000010429f680, parent=0x00000001042a16f0) at ogre_render_window_impl.cpp:95
    frame #16: 0x00000001016cd439 librviz_rendering.dylib`rviz_rendering::RenderWindow::RenderWindow(this=0x00000001042a16f0, parent=0x0000000000000000) at render_window.cpp:65
    frame #17: 0x00000001016cd55d librviz_rendering.dylib`rviz_rendering::RenderWindow::RenderWindow(this=0x00000001042a16f0, parent=0x0000000000000000) at render_window.cpp:67
    frame #18: 0x00000001000ee356 rviz2_app`rviz_common::RenderPanel::RenderPanel(this=0x000000010429f5b0, parent=0x000000010429f760) at render_panel.cpp:74
    frame #19: 0x00000001000eed0d rviz2_app`rviz_common::RenderPanel::RenderPanel(this=0x000000010429f5b0, parent=0x000000010429f760) at render_panel.cpp:76
    frame #20: 0x00000001001d2d04 rviz2_app`rviz_common::VisualizationFrame::initialize(this=0x0000000104296a80, display_config_file=0x00007ffeefbff3a0) at visualization_frame.cpp:275
    frame #21: 0x0000000100202179 rviz2_app`rviz_common::VisualizerApp::init(this=0x00007ffeefbff5e8, argc=1, argv=0x00007ffeefbff768) at visualizer_app.cpp:285
    frame #22: 0x0000000100012559 rviz2_app`main(argc=1, argv=0x00007ffeefbff768) at main.cpp:61
    frame #23: 0x00007fff69466145 libdyld.dylib`start + 1
    frame #24: 0x00007fff69466145 libdyld.dylib`start + 1
(lldb)
```

I might merge this and revert the last commit, assuming that would fix the problem.

File: rviz_ogre_vendor/CMakeLists.txt [hunk 2]
```diff
@@ -16,7 +16,8 @@ macro(build_ogre)
   list(APPEND extra_cmake_args "-DOGRE_BUILD_RENDERSYSTEM_D3D11:BOOL=OFF")
   list(APPEND extra_cmake_args "-DOGRE_BUILD_RENDERSYSTEM_D3D9:BOOL=OFF")
   if(APPLE)
-    list(APPEND extra_cmake_args "-DCMAKE_OSX_ARCHITECTURES='x86_64'")
+    list(APPEND extra_cmake_args "-DCMAKE_OSX_ARCHITECTURES='x86_64'"
+      "-DCMAKE_CXX_FLAGS=-std=c++14 -stdlib=libc++")
```
Comment (wjwwood) [inline rviz_ogre_vendor/CMakeLists.txt:20]: Hmm, this is strange too, the `-stdlib` should not be needed and the `-std=c++14` is set just a few lines above already.

File: rviz_rendering/src/rviz_rendering/render_system.cpp [hunk 3]
```diff
@@ -408,10 +408,7 @@ RenderSystem::makeRenderWindow(
 
   params["currentGLContext"] = Ogre::String("false");
 
-  // This line is needed to receive mouse events on Windows, but results in a crash on OS X
-#ifdef _WIN32
   params["externalWindowHandle"] = Ogre::StringConverter::toString(window_id);
-#endif
```
Comment (wjwwood) [inline rviz_rendering/src/rviz_rendering/render_system.cpp:414]: Btw, this changes the behavior on Linux as well as macOS.

File: rviz_ogre_vendor/CMakeLists.txt (hunk 2 above)
Comment (wjwwood) [reply rviz_ogre_vendor/CMakeLists.txt:20]: Yet... it does seem needed, I think we should report this upstream. They must be setting the `-stdlib` to something wrong unless we override it or something. There's no other reason that this file `OgreProperty.h` would be able to do `#include <functional>` but not use `std::function` later.

Comment (wjwwood) [reply rviz_ogre_vendor/CMakeLists.txt:20]: Might be something my patch was fixing before:

https://github.com/ros2/rviz/pull/127/files#diff-d20b3504260c943b0679a45f137ff635L28

File: rviz_rendering/src/rviz_rendering/render_system.cpp (hunk 3 above)
Comment (wjwwood) [reply rviz_rendering/src/rviz_rendering/render_system.cpp:414]: If I revert this particular change my rviz starts working again... I don't know if this is specific to my machine or not.

**Comment by wjwwood** [comment]
I'm going to merge this and fix it up in the use pluginlib branch.

File: rviz_rendering/src/rviz_rendering/render_system.cpp (hunk 3 above)
Comment (greimela-si) [reply rviz_rendering/src/rviz_rendering/render_system.cpp:414]: > Btw, this changes the behavior on Linux as well as macOS.

In our experience the `externalWindowHandle` param is ignored on linux and does not influence the behavior.

**Pull request merged**